### PR TITLE
Features/451 isclose

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

numpy.isclose() gets called by naive_bayes.GaussianNB . This is the heat version that works on distributed tensors as well. Returns a boolean tensor or (if the input is scalar) a single boolean value.

Issue/s resolved: #451 

## Changes proposed:
- Added isclose() to module logical, and as a method to DNDarray
- Moved sanitation of input values/tensors out of logical.allclose() into a helper function logical._sanitize_close_input(). Both allclose and isclose now call _sanitize_close_input()

## Known issues! 
#457  logical._sanitate_close_input should check shapes, and broadcast before redistributing
#458 get logical.isclose to work with unbalanced tensors

## Type of change

Remove irrelevant options:
- New feature (non-breaking change which adds functionality)
- Documentation update

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)

#### Does this change modify the behaviour of other functions? If so, which?
 no

### CODE REVIEW

File: heat/core/logical.py [hunk 1]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
     tensor([ False, False,  True])
     """
     return operations.__binary_op(torch.logical_xor, t1, t2)
+
+
+def _sanitize_close_input(x, y):
```
Comment (coquelin77) [inline heat/core/logical.py]: 2 underscores at the start of internal functions

File: heat/core/logical.py [hunk 2]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
...
+    if np.isscalar(x):
+        try:
+            if isinstance(y, dndarray.DNDarray):
+                x = factories.array(float(x), device=y.device)
+            else:
+                x = factories.array(float(x))
```
Comment (coquelin77) [inline heat/core/logical.py]: device argument

File: heat/core/logical.py [hunk 3]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
...
+        raise TypeError(
+            "Only tensors and numeric scalars are supported, but input was {}".format(type(x))
+        )
+
+    if np.isscalar(y):
+        try:
+            if isinstance(x, dndarray.DNDarray):
+                y = factories.array(float(y), device=x.device)
+            else:
+                y = factories.array(float(y))
```
Comment (coquelin77) [inline heat/core/logical.py]: device

File: heat/core/logical.py [hunk 4]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
...
+    # If both x and y are split, but along different axes, y is redistributed to be split along the same axis as x
+    # TODO: needs broadcasting first depending on shapes
+    elif (x.split is not None) and (y.split is not None) and (x.split != y.split):
+        t1 = x.copy()
+        t2 = manipulations.resplit(y, axis=x.split)
+
+    else:
+        t1 = x.copy()
+        t2 = y.copy()
+
```
Comment (coquelin77) [inline heat/core/logical.py]: there are lots of copy calls here. do we need to copy the data? do any of the 'close' functions overwrite the data?

File: heat/core/logical.py [hunk 5]
```diff
@@ -214,6 +180,47 @@ def local_any(t, *args, **kwargs):
     )
 
 
+def isclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
+    """
+    Parameters:
+    -----------
+
```
Comment (coquelin77) [inline heat/core/logical.py]: remove extra blank line between  '--------' and text

File: heat/core/logical.py [hunk 6]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
...
+                x = factories.array(float(x))
+        except (ValueError, TypeError):
+            raise TypeError("Data type not supported, input was {}".format(type(x)))
+
+    elif not isinstance(x, dndarray.DNDarray):
+        raise TypeError(
+            "Only tensors and numeric scalars are supported, but input was {}".format(type(x))
+        )
+
+    if np.isscalar(y):
```
Comment (Markus-Goetz) [inline heat/core/logical.py]: Code duplication, factor out in a separate function

File: heat/core/dndarray.py [hunk 7]
```diff
@@ -928,6 +928,28 @@ def __eq__(self, other):
         """
         return relational.eq(self, other)
 
+    def isclose(self, other, rtol=1e-05, atol=1e-08, equal_nan=False):
+        """
+        Parameters:
+        -----------
+
```
Comment (coquelin77) [inline heat/core/dndarray.py]: extra blank lines here between '------' and text

File: heat/core/logical.py [hunk 8]
```diff
@@ -307,3 +314,55 @@ def logical_xor(t1, t2):
...
+            raise TypeError("Data type not supported, input was {}".format(type(y)))
+
+    elif not isinstance(y, dndarray.DNDarray):
+        raise TypeError(
+            "Only tensors and numeric scalars are supported, but input was {}".format(type(y))
+        )
+
+    # Do redistribution out-of-place
+    # If only one of the tensors is distributed, unsplit/gather it
+    if (x.split is not None) and (y.split is None):
```
Comment (Markus-Goetz) [inline heat/core/logical.py]: Are these parenthesis really pep-8 conform? Do you have the black commit hook installed?

File: heat/core/logical.py (hunk 1 above)
Comment (Markus-Goetz) [inline heat/core/logical.py]: Use double leading underscores for internal functions (yes we might have violated this in the past)

Comment (ClaudiaComito) [reply heat/core/logical.py]: Thanks, done.

File: heat/core/logical.py (hunk 3 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: I've rewritten this entirely.

File: heat/core/logical.py (hunk 5 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: Done

File: heat/core/dndarray.py (hunk 7 above)
Comment (ClaudiaComito) [reply heat/core/dndarray.py]: Done

File: heat/core/logical.py (hunk 6 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: Right, thanks, I've rewritten this part

File: heat/core/logical.py (hunk 8 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: haha, I missed this. I do have the black commit hook installed as far as I can tell, and it hasn't complained, but I've removed the parentheses.

File: heat/core/logical.py (hunk 1 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: Done thanks

File: heat/core/logical.py (hunk 4 above)
Comment (ClaudiaComito) [reply heat/core/logical.py]: They don't overwrite the data, but they do resplit the tensors in some cases. 

The last conditional statement doesn't need to copy though, I don't think. I've changed it to simply return x, y.

File: heat/core/logical.py [hunk 9]
```diff
@@ -307,3 +312,49 @@ def logical_xor(t1, t2):
...
+        return x
+
+    x = sanitize_input_type(x, y)
+    y = sanitize_input_type(y, x)
+
+    # Do redistribution out-of-place
+    # If only one of the tensors is distributed, unsplit/gather it
+    if x.split is not None and y.split is None:
+        t1 = manipulations.resplit(x, axis=None)
+        t2 = y.copy()
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/logical.py]: you're right, I condensed it a bit more and removed the unnecessary copy calls

**Pull request merged**